pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -0,0 +1,686 @@
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, nil
+				}
+				state.rListerMock.GetFunc = func(_, _ string) (*rbacv1.Role, error) {
+					return nil, apierrors.NewNotFound(schema.GroupResource{
+						Group:    normanv1.RoleGroupVersionKind.Group,
+						Resource: normanv1.RoleGroupVersionResource.Resource,
+					}, "")
+				}
+				state.rClientMock.CreateFunc = func(role *rbacv1.Role) (*rbacv1.Role, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  true,
+		},
+		{
+			name: "create roles successfully",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_, _ string) (*corev1.Namespace, error) {
+					return nil, nil
+				}
+				state.rListerMock.GetFunc = func(namespace string, name string) (*rbacv1.Role, error) {
+					return nil, apierrors.NewNotFound(schema.GroupResource{
+						Group:    normanv1.RoleGroupVersionKind.Group,
+						Resource: normanv1.RoleGroupVersionResource.Resource,
+					}, "")
+				}
+				state.rClientMock.CreateFunc = func(role *rbacv1.Role) (*rbacv1.Role, error) {
+					state.stateChanges.createdRoles[role.Name] = role
+					role.UID = ""
+					return role, nil
+				}
+			},
+			stateAssertions: func(gtsc grTestStateChanges) {
+				require.Len(gtsc.t, gtsc.createdRoles, 2)
+
+				role, ok := gtsc.createdRoles["namespacedRulesGR-namespace1"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, "namespace1", role.Namespace)
+				require.Equal(gtsc.t, "namespacedRulesGR", role.Labels[grOwnerLabel])
+				require.Len(gtsc.t, role.Rules, 2)
+				require.Equal(gtsc.t, readPodPolicyRule, role.Rules[0])
+				require.Equal(gtsc.t, readConfigPolicyRule, role.Rules[1])
+				require.Equal(gtsc.t, namespacedRulesOwnerRef, role.OwnerReferences[0])
+
+				role, ok = gtsc.createdRoles["namespacedRulesGR-namespace2"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, "namespace2", role.Namespace)
+				require.Equal(gtsc.t, "namespacedRulesGR", role.Labels[grOwnerLabel])
+				require.Len(gtsc.t, role.Rules, 2)
+				require.Equal(gtsc.t, adminPodPolicyRule, role.Rules[0])
+				require.Equal(gtsc.t, readPodPolicyRule, role.Rules[1])
+				require.Equal(gtsc.t, namespacedRulesOwnerRef, role.OwnerReferences[0])
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  false,
+		},
+		{
+			name: "some roles have errors",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsListerMock.GetFunc = func(_ string, name string) (*corev1.Namespace, error) {
+					if name == "namespace2" {
+						return nil, fmt.Errorf("not found")
+					}
+					return nil, nil
```

Comment:
It's not clear to me what we're trying to assert here.

The reconciler loops over each namespaces (unordered due to map) and tries to reconcile the rules for those namespaces. If any of them fail, it still keeps going.

This test here only check that the first namespace (in most cases, again, unordered map) is created.
I think it would be best to verify that even though an error occur, other namespaces are also reconciled properly after.

Something like:
- ns1 error
- ns2 success

What do you think?

Reply:
You are absolutely correct. That was what I was intending to test but like you said, an unordered map doesn't really guarantee that one goes before the others. I'll still swap them like you suggested

Reply:
I think what you could do instead of just swapping would be to have something like:

```go
state.nsListerMock.GetFunc = func(_ string, name string) (*corev1.Namespace, error) {
					if state.getCount == 0 {
						state.getCount += 1
						return nil, fmt.Errorf("not found")
					}
					return nil, nil
```
so you can be sure that the loop fails for the first item.

Of course now your assertions will depend on which role was successfully created.. What do you think of that?